source/Cosmos.HAL2/Audio/AudioBuffer.cs

```diff
@@ -0,0 +1,234 @@
+﻿using Cosmos.Core;
+using Cosmos.HAL.Audio;
+using System;
+
+namespace Cosmos.HAL.Audio {
+    /// <summary>
+    /// Represents a buffer of channel-interleaved audio samples.
+    /// </summary>
+    public class AudioBuffer {
+        int size;
+        byte[] buffer;
+        public readonly SampleFormat format;
+        readonly int sampleAmount;
+
+        /// <summary>
+        /// Creates an empty <see cref="AudioBuffer"/>.
+        /// </summary>
+        /// <param name="size">The size of the buffer.</param>
+        /// <param name="format">The target format of the audio buffer.</param>
+        public AudioBuffer(int size, SampleFormat format)
+        {
+            this.format = format;
+            this.size = size;
+            buffer = new byte[format.size * Size];
+            sampleAmount = size * format.channels;
+        }
+
+        /// <summary>
+        /// Provides access to the raw buffer data.
+        /// </summary>
+        public byte[] RawData => buffer;
+
+        /// <summary>
+        /// The size of the audio buffer, in audio samples. Setting this property
+        /// to a different value causes the audio buffer to flush.
+        /// </summary>
+        public int Size {
+            get => size;
+            set {
+                if(value != size)
+                {
+                    size = value;
+                    buffer = new byte[format.size * Size];
+                }
+            }
+        }
+
+        /// <summary>
+        /// The amount of sample values the buffer contains.
+        /// </summary>
+        public int SampleAmount => sampleAmount;
+
+        /// <summary>
+        /// Flushes (empties) the audio buffer, setting all of its sample data to zero.
+        /// </summary>
+        public void Flush()
+        {
+            int calculatedSize = format.size * Size;
+
+            if(buffer != null && buffer.Length == calculatedSize) {
+                // The buffer already exists and the size has not changed -
+                // set all of the bytes of the buffer to 0.
+                MemoryOperations.Fill(buffer, 0);
+            } else {
```

Comment:
Can this case actually happen where the size is different? From the function name i would expect the size to already be correct.

Reply:
Oops, yeah, sorry! This is a remnant of an old version of the API I was working on where the buffer was actually created within this method. I'll create a commit that removes this case ASAP!